Update agent setup instructions (#94)
neo4j/docs-ops-manager: modules/ROOT/pages/addition/agent-installation/self-registered.adoc

= Self-registering agent
:description: this page describes NOM agent self-registering setup.

> Agent self registration omits manually aquiring initial configuration from NOM UI as described in xref:./manual.adoc#register[Manually registering agent] and instead relies on additional agent configuration on Neo4j instance host to obtain initial configuration from NOM server either through mutual authentication or manual authorization of the agent from UI after installation.

Reviewer: ```suggestion
Agent self-registration omits manually acquiring initial configuration from NOM UI as described in xref:./manual.adoc#register[Manually registering agent].
Instead minimal generic configuration about the NOM server is provided and the additional agent specific configuration is obtained from the NOM server after authorization. 
Authorization is obtained either through mutual authentication or manual authorization of the agent from UI. 
```

[[configure]]
== Configuration for agent self registration with manual approval
Following are the required environment variables to be configured to enable self-registration of NOM agent:

[cols="<,<,<",options="header"]
|===
| Variable
| Description
| Example

| `CONFIG_SERVER_GRPC_ADDRESS`
| Server GRPC Address
| server:9090

| `CONFIG_SERVER_HTTP_ADDRESS`
| Server HTTP address
| +++https://server:8080+++
|===

With only the above configuration, agent attempts to connect to NOM server for self-registration which results in agent going into *Standby mode* until authorized.

IMPORTANT: *Standby mode* - this describes the runtime state of a self registering agent which is not authorized yet to monitor an instance. In this mode agent keeps checking with NOM server for approval status at every preset time interval (__default 30 seconds__).

An agent in standby mode shows up in the list of agents in NOM UI. To enable the agent to continue it's normal execution, agent needs to be approved from NOM UI as shown below:

image::agents.png[width=800]

To approve the agent click on `...` agent action icon and click `Approve`:

image::agents.png[width=800]

> An edit view to optionally update agent name or description will show as below:

Reviewer: Does this have to be done before approval to be correctly persisted or will it still be persisted if it's done later?

If it can be done later - I think we should move this step to the end to keep the flow simple.

Author: it can be done anytime

Reviewer 2: I think we should remove this step from the instruction in the interest of brevity - would think that the ability to change the name would be obvious enough.

NOTE: an unauthorized self-registering agent will have default name derived from it's host information. It's recommended to update this name or use the optional configuration described <<agent-meta-data,here>>.

image::agent-new.png[width=800]

> Once the agent is approved, by optionally updating it's name and description, agent will receive initial configuration from NOM server which it persists on the instance host in a file named `nom-agent-config.yaml`.

Reviewer: ```suggestion
Once the agent is approved, the agent will receive initial configuration from NOM server which it persists on the instance host in a file named `nom-agent-config.yaml`.
```

WARNING: Location of where agent saves it's initial configuration must be of persistent type else agent needs to be reregistered.

Following optional configuration can be used to provide a persistent location for agent intial configuration:

[cols="<,<,<,<",options="header"]
|===
| Variable
| Description
| Example
| Default

| `CONFIG_AGENT_CONFIG_PATH`
| Peristent path to a file on Neo4j instance host
| "file://path/to/"

> | by default agent tries to save initial configuration in path `NEO4J_CONF` if set or `conf` folder under `NEO4J_HOME` if set, else the file is saved under `.nom` folder in user home directory.

Reviewer: What if no instances are configured? 
What if multiple instances are configured?

Author: No instances would result in config being saved to user home directory. for multiple instances, since agent looks for current env for config folder path, it could only be one value or no value per host right?

> | by default agent tries to save initial configuration in path `NEO4J_CONF` if set or `conf` folder under `NEO4J_HOME` if set, else the file is saved under `.nom` folder in user home directory.

Reviewer: ```suggestion
| `NEO4J_CONF` if set or `conf` folder under `NEO4J_HOME` if set, else `.nom` folder in user home directory.
```

|===

> Apart from above configuration, following optional configuration can be provided as agent meta data:

Reviewer: ```suggestion
Agent meta-data can be optionally specified using these configuration parameters:
```

[[agent-meta-data]]
[cols="<,<,<",options="header"]
|===
| Variable
| Description
| Example

> | `CONFIG_AGENT_NAME`

Reviewer: Can this additional meta-data be provided in the mtls set up too?

Author: yes

| Optional name for agent to easily differentiate among self-registered agents
| home-db-agent

| `CONFIG_AGENT_DESCRIPTION`
| Optional description for agent to easily differentiate among self-registered agents
| An agent to monitor home db
|===

== Configuration for agent self registration with mutual authentication
Following are the required environment variables to be configured to enable self-registration of NOM agent with mutual authentication:

[cols="<,<,<",options="header"]
|===
| Variable
| Description
| Example

| `CONFIG_SERVER_GRPC_ADDRESS`
| Server GRPC Address
| server:9090

| `CONFIG_SERVER_HTTP_ADDRESS`
| Server HTTP address
| +++https://server:8080+++

| `CONFIG_CLIENT_CERT`
| PEM encoded Agent certificate for mutual TLS
| /path/to/a/pem/file`

| `CONFIG_CLIENT_KEY`
| PEM encoded Agent key for mutual TLS
| /path/to/a/pem/file`
|===

IMPORTANT: *In addition to above configuration, NOM server also needs to be configured to trust the agent certificates as described xref:/installation/server.adoc#config_ref[here]*

=== Self signed certificates for agent

[IMPORTANT]
====

> Self-signed certificates are not recommended to be used in production environments.

Reviewer 2: Is there a reason we're not recommending self-signed certs for mTLS in production? I get it with HTTPS certs that you'd want to have a proper CA signed cert, but for self-registration it shouldn't matter as long as the key is kept safe. Unless I'm missing something?

Author: You are right about the agent keys being a maintenance overhead, we can still suggest it though

+
[source, terminal, role=noheader]
----
agent version
----
+
Which should output:
+
[source, terminal, role=noheader]
----
<agent name> <agent version> <agent revision>
----

* Make sure the agent version is equal to that of the NOM server.

[[configure]]
== Configuration for agent self registration with manual approval
Following are the required environment variables to be configured to enable self-registration of NOM agent:

[cols="<,<,<",options="header"]
|===
| Variable
| Description
| Example

| `CONFIG_SERVER_GRPC_ADDRESS`
| Server GRPC Address
| server:9090

| `CONFIG_SERVER_HTTP_ADDRESS`
| Server HTTP address (** Http address should include protocol scheme **)
| +++https://server:8080+++
|===

With only the above configuration, agent attempts to connect to NOM server for self-registration which results in agent going into *Standby mode* until authorized.

IMPORTANT: *Standby mode* - this describes the runtime state of a self registering agent which is not authorized yet to monitor an instance. In this mode agent keeps checking with NOM server for approval status at every preset time interval (__default 30 seconds__).

> WARNING: Location of where agent saves it's initial configuration must be of persistent type else agent needs to be reregistered.

Reviewer: ```suggestion
WARNING: agent config location must be of persistent type. 
```

WARNING: Location of where agent saves it's initial configuration must be of persistent type else agent needs to be reregistered.

> Following optional configuration can be used to provide a persistent location for agent to save it's initial configuration:

Reviewer 2: This sort of makes it sound like the initial configuration will not be persisted unless the optional parameter is used 🤔

> Following optional configuration can be used to provide a persistent location for agent to save it's initial configuration:

Reviewer 2: ```suggestion
The following optional configuration can be used to specify the location for agent to save its initial configuration:
```

[[running-agent]]
== Run agent

A self-registering agent is preferably run in console mode only and an additional command line option should be provided to enable self-registration.

* Run an agent in console mode:
+
All configuration values for the agent should be set as environment variables before starting the agent
+
** Command:
+
[source, terminal, role=noheader]
----
agent console -s
----
+
Or
+
[source, terminal, role=noheader]
----
agent console --self-register
----

[[verify]]
== Verify agent setup
Ensure agent has contacted NOM server, is online and is reporting DBMS(s) correctly.

. Return to Agents listing in global settings
+
image::agents.png[width=800]
. An agent in standby mode shows up in the list of agents in NOM UI with `Unauthorized` status. To enable the agent to continue it's normal execution, agent needs to be approved from NOM UI as shown below:
+
image::agent-unauthorized.png[width=800]
. To approve the agent click on `...` agent action icon and click `Approve`:
+
image::agent-action-menu.png[width=800]
. An edit view to optionally update agent name or description will show as below:
+

> NOTE: an unauthorized self-registering agent will have default name derived from it's host information. It's recommended to update this name or use the optional configuration described <<agent-meta-data,here>>.

Reviewer 2: ```suggestion
NOTE: An unauthorized self-registering agent will have a default name derived from its host information. It's recommended to update this name or use the optional configuration described <<agent-meta-data,here>>.
```

+
image::agent-approve.png[width=800]
. Once the agent is approved, the agent will receive initial configuration from NOM server which it persists on the instance host in a file named `nom-agent-config.yaml`.

> . Upon approval agent status changes to `Offline` until agent connects to Server with initial configuration.

Reviewer 2: ```suggestion
. Upon approval, the agent status changes to `Offline` until the agent connects to the server with the initial configuration.
```
